%To understand who this system is built for and what it needs to do, the following chapter will explain the target group and its needs.

The \appName\ system was designed with a specific target group in mind: Epigenitic researchers. This chapter will explain the needs of these users, the problems they faced before this system was provided and the requirements that were collected and taken into account for this project.

\section{Target group}

The main group that wants to utilize the \appName\ system is the researchers at \term{Epigenetic Cooperation Norrland (EpiCoN)}. They are conducting experiments on \term{DNA} and proteins in order to see where certain proteins bind to \term{DNA} strings. This information, combined with knowledge about the location of genes within a given genome, will give the researchers a hint of which proteins are active in enabling and disabling genes. Such experiments yield important results for the study of how cells remember which genes should be enabled after cell division.

Previous to the \appName\ project the raw data files retrieved from experiments were manually processed by the researchers using inefficient \term{Perl} script. This process also involved using \term{Bowtie}\cite{langmead2009ultrafast}, a program used to unscramble the \term{DNA} data and \term{LiftOver}\cite{LIFTOVER}, a program used to adjusting results to a different genome release.

The researchers at \term{EpiCoN} have varying computer skills. While they all have basic computer knowledge, not all are familiar with more advanced computer tasks such as running scripts at command line level. This resulted in some researchers being dependent on the ones who could process the raw data. At \term{EpiCon} the researcher that had the knowledge to use all the scripts and software ended up performing these time consuming tasks for other researchers.

From time to time students are interested in working with the data, however they are not as trusted as the regular researchers and are not allowed to manipulate the data, only look at it and do analyses. The researchers at \term{EpiCoN} are a diverse group encompassing many different nationalities and therefore communicate in English.

%Students at \term{Ume√• University} sometimes have the need to look at research done by \term{EpiCoN}. These students doesn't have the same trust to handle the research as the main staff at \term{EpiCoN}. The researchers at \term{EpiCoN} are a diverse group of many nationalities. They are therefore communicating in english.

%Their main field of knowledge lies within epigenetics. It is an international group with many different native languages, which means that they mainly use english to communicate.

%Since biologists have varying experience with software and software development, the system will be designed to be as user friendly as possible. The international language of science is english, therefore the entire product shall be in english (both the product and its documentation). Since target audience has different knowledge in biology, the product and its functions shall be simple, self explanatory and have an interface. The biology students are not as trusted as the standard users and therefore their data access will be restricted in some way. The system might also have some administrators. An administrator primarily handles new users and user rights. If the system expands the administrator may also have to maintain the system. For starters some of the biologists with more computer knowledge will be given authority to handle users and user rights.

\section{Client needs}
The target group of researchers from \term{EpiCoN} needs a system to structure the large amount of genetic data that they use daily. The data is used for research of cell memory, how cells remember which sequences in the \term{DNA} that should be active after cell division. The requirements of the system are described below and were collected in a number of \term{user stories} each of which describe a desired function from the end users perspective. All \term{user stories} are presented in \refer{chap:userstories}, when discussed below the title of the \term{user story} will be stated.

There are mainly three data types used for this research and that the system should handle: \term{raw}, \term{profile} and \term{region} data. \term{Raw} data is the raw output from an experiment and isn't usually analysed directly due to the large nature of the files. It is processed to so called \term{profile} data. \term{Profile} data describes the amount of reads found for every base--pair in an organism's genome. \term{Region} data is further processed \term{profile} data consisting of the regions where every base--pair's read strength is above a given threshold and fault tolerance. The region gets a value based on the average of the base-pair reads for the given region.

\subsection{Storage}

When conducting experiments the researchers treat DNA with a certain protein that binds to certain positions. The DNA-strings are then broken into pieces and the bound areas are collected. In this step \term{raw} data is collected on what these pieces look like on a base--pair level. This data is the foundation for further research and must be stored securely and in a structured manner. A system that automates the archiving of \term{raw} data to a shared location is therefore desired. This requirement was specified in the \term{user stories} ''Single Upload'' (see \refer{fig:target_upload}) and ''Single Download'' (see \refer{fig:target_download}).

\begin{figure}[h]
\userstory{Single upload}{To store a single data file \\ the researchers \\ want to be able to upload a specific file.}
\caption{User story for uploading a file to the central database}
\label{fig:target_upload}
\end{figure}

\begin{figure}[h]
\userstory{Single download}{To scrutinize a single data file \\ the researchers \\ want to be able to download a specific file.}
\caption{User story for downloading a file from the central database}
\label{fig:target_download}
\end{figure}

However researchers do not always have to generate the \term{raw} data in order to conduct research. When results are published in scientific articles the \term{raw} data from the experiments are often also provided. One location where these \term{raw} data files can be published is the \term{GEO} (\term{Gene Expression Omnibus}) database. A desire to be able to initialise a download of \term{raw} data to \appName\ from this source was also expressed and written up in the \term{user story} ``Download from GEO Database''.

It is not possible/highly impractical to gain knowledge of an experiment by viewing the characters that make up the resulting \term{raw} files. In order to see information about an experiment and the resulting files, the researchers must be able to annotate them upon entry to the database. There are several \term{user stories} to do with the annotation of experiments/files and the manipulation of the possible annotation fields. An example of such a \term{user story} is ``Annotation'' (see \refer{fig:target_annotation}).

\begin{figure}[h]
\userstory{Annotation}{To structure the data files \\ the researchers \\ want to be able to annotate the data files.}
\caption{User story for annotation of files}
\label{fig:target_annotation}
\end{figure}

As they fill the database with their data, the researchers want to make sure that it is kept safe. They therefore want to have an authorisation system to protect the data from unauthorised access. Another risk that has to be mitigated is the risk of hardware failure. These concerns are expressed through the \term{user stories} ''Password Protected'' and ''Backup'' respectively (see \refer{fig:target_password} and \refer{fig:target_backup}).

\begin{figure}[h]
\userstory{Password protected}{To protect the database from unauthorized use \\ the researchers \\ want the application to be password protected.}
\caption{User story for uploading a file to the central database}
\label{fig:target_password}
\end{figure}

\begin{figure}[h]
\userstory{Backup}{To prevent loss of data \\ the researchers \\ want the data to be backed up.}
\caption{User Story about the protection from hardware failures}
\label{fig:target_backup}
\end{figure}

The experiment methods result in a data set that requires processing in order to conduct meaningful analysis. The researchers have themselves composed various scripts, which when combined with the  \term{BowTie} algorithm yeild usable processed \term{profile} data. In this format the \term{DNA} pieces are ordered and mapped to the real \term{DNA} string. The result is data for how the particular protein binds certain parts of the \term{DNA} string. In order to be useful the system should provide a way of automatically carry out these analysis steps required to convert \term{raw} data to \term{profile} data (see \refer{fig:target_r2p}). 

\begin{figure}[h]
\userstory{Raw to profile}{To be able to analyze \\ the researchers \\ want to process raw data to profile data.}
\caption{Convert from raw data to profile data format}
\label{fig:target_r2p}
\end{figure}

After processing, the resulting \term{profile} data files should be annotated and saved in the database alongside the original raw data files. It is important that they are traceable to the original \term{raw} data files and that the parameters used in their creation are saved so that the analysis process can be recreated and confirmed (see \refer{fig:target_trace}).

\begin{figure}[h]
\userstory{File traceabillity}{To be able to access the underlying raw data or profile data \\ the researchers \\ want the raw data files to be traceable from profile files and the profile files to be traceable from the region data (if available) when the files have been generated on the server.}
\caption{Traceability \term{user story}}
\label{fig:target_trace}
\end{figure}

% Most importantly the researchers want a central secure location to store their genome data in a structured  way so that it is easy to find the data the researchers are looking for when doing research. By having a central location for genome data the researchers can more easily collaborate than they do today. The researchers want the database to be locked down from outside access and that the data is stored in a secure way to avoid loss of data (''Password protected'', ''Backup''). Data integrity is of great importance, data must not change or become corrupt in the database.

%\subsection{File formats}
%As different software uses different file formats for genome data (''Convert common file formats''), and the tools for converting files between these file formats are not very user friendly. To do analyses on the data it must be processed from raw its raw form, this conversion is a very time consuming process. The researchers want this work to be carried out by the system (''\term{Raw} to \term{Profile}'', ''\term{Profile} to \term{Region}'').

\subsection{Genome release}
As new discoveries are made in the area, new standards for the order of the base pairs in the \term{DNA} string are set. This results in a new \term{Genome Release} for a specific species which is a set of files specifying the new standard. This also affects the order used in the creation of \term{profile} data files. The researchers need a way to convert \term{profile} data files between these releases (see \refer{fig:target_genome}). It is not uncommon for errors in a new release to be discovered after publication. It is therefore important to keep data files that are compatible with an older release for some time after a new release is published.

\begin{figure}[h]
\userstory{Convert genome release}{To easier handle files \\ the researchers \\ want to convert files between genome releases (\term{LiftOver}).}
\caption{To convert between different genome releases}
\label{fig:target_genome}
\end{figure}

\subsection{Analysis}
The researchers want to do some analyses on the server. Some analyses is to combine regions using logical arithmetic and in such a way construct new regions (''Combine regions''). It should also be possible to create new regions from a reference point (''Create region subset''). Another interesting analysis is overlap analysis which shows how much a number of genome experiments overlap(''Overlap analysis'').

\subsection{Visualization}
To be able to review analysis results, the researchers would like a graphical presentation of the results(''Plot overlap analysis'', ''Plot average regions''). They also want to be able to use the \term{Integrated Genome Browser} (IGB) software to view results (''IGB Session''). They therefore want to be able to download a session-file to immediately start the browser.